Update data.js to handle fieldSelector/labelSelector

benjaminapetersen opened the pull request:
[Original PR against console_1.3](https://github.com/jwforres/origin/pull/40) w/conversation.

@jwforres @spadgett @liggitt

benjaminapetersen commented on assets/app/scripts/filters/resources.js:
```diff
@@ -15,27 +15,35 @@ angular.module('openshiftConsole')
       return annotationMap[annotationKey] || null;
     };
   })
+  .filter('labelName', function() {
```
Open question on previous PR, copy over the map from `annotationName` filter?

jwforres replied:
no i think some of them are actually inconsistent between the label and the annotation because of legacy label keys, we'll add them as we need them

benjaminapetersen replied:
sounds good

benjaminapetersen commented:
Getting a fork_ami rolling for this.

jwforres commented on assets/app/scripts/services/data.js:
```diff
@@ -101,19 +105,20 @@ angular.module('openshiftConsole')
     callbacks.add(callback);
 
-    if (this._watchInFlight(resource, context) && this._resourceVersion(resource, context)) {
+    if (this._watchInFlight(key) && this._resourceVersion(key)) {
```
` && (!!this._data(key))`

jwforres commented on assets/app/scripts/services/data.js:
```diff
@@ -299,21 +304,21 @@ angular.module('openshiftConsole')
       }, 0);
     }
-    else if (!force && this._watchInFlight(resource, context) && this._resourceVersion(resource, context)) {
+    else if (!force && this._watchInFlight(key) && this._resourceVersion(key)) {
```
` && (!!this._data(key))`

jwforres replied:
with all the instances of this, and since you've already updated everything to just take key, i think you should go ahead and add that `isCached(key)` you mentioned on my pr, makes more sense to do it here and make sure you get all of the instances of this

benjaminapetersen replied:
Agree, the repetition makes me nervous.  Will update!

benjaminapetersen replied:
Looking at `_isResourceCached` as well to make sure there is a logical reason for both.

```JavaScript
  var CACHED_RESOURCE = {
    imagestreamimages: true
  };

  DataService.prototype._isResourceCached = function(resource) {
    return !!CACHED_RESOURCE[resource.resource];
  };
```

benjaminapetersen replied:
- [ ] add `_isCached(key)` and ensure it is clearly different than `_isResourceCached(resource);`

jwforres replied:
_isResourceCached is whether a particular type of resource should be put
into the cache, i think there was one we didnt want to cache ever

On Fri, Apr 29, 2016 at 3:47 PM, Ben Petersen <notifications@github.com>
wrote:

> In assets/app/scripts/services/data.js
> <https://github.com/openshift/origin/pull/8676#discussion_r61634891>:
>
> >
> >      // If this is a cached resource (immutable resources only), ignore the force parameter
> >      if (this._isResourceCached(resource) && existingData && existingData.by('metadata.name')[name]) {
> >        $timeout(function() {
> >          deferred.resolve(existingData.by('metadata.name')[name]);
> >        }, 0);
> >      }
> > -    else if (!force && this._watchInFlight(resource, context) && this._resourceVersion(resource, context)) {
> > +    else if (!force && this._watchInFlight(key) && this._resourceVersion(key)) {
>
>
>    - add _isCached(key) and ensure it is clearly different than
>    _isResourceCached(resource);
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8676/files/cd51fd4ee136424df96ea874ef68e339fef98098#r61634891>
>

benjaminapetersen replied:
I see this comment:
```JavaScript
// If this is a cached resource (immutable resources only), ignore the force parameter
if (this._isResourceCached(resource) && existingData && existingData.by('metadata.name')[name]) {
```
In my snippet above, the only item in the hash is `imagestreamimages`, so that is the only resource that ever returns `true`.  

Reading it this way implies that `imagestreamimages` would be the only thing we ever cache, but I think we mean the opposite.

spadgett replied:
I read the comment as saying the cached version of the resource is always current since it's immutable. (+1 to changing the name.)

jwforres commented on assets/app/scripts/services/data.js:
```diff
@@ -815,59 +807,74 @@ DataService.prototype.createStream = function(resource, name, context, opts, isR
+    return _.reduce(
+            keys,
+            function(result, key, i) {
+              return (i < (keys.length-1)) ?
```
trying to decide whether i'd rather see what you've got or this:
```
return result + key + '=' + encodeURIComponent(params[key]) +
       (i < (keys.length-1)) ? '&' : ''
```

benjaminapetersen replied:
I like that, eliminates most of the repeated line, I'll update.

jwforres commented on assets/app/scripts/services/data.js:
```diff
@@ -815,59 +807,74 @@ DataService.prototype.createStream = function(resource, name, context, opts, isR
+    var ns = context && context.namespace ||
+             _.get(context, 'project.metadata.name') ||
+             context.projectName;
+    return this._urlForResource(resource, name, context, null, angular.extend({}, {}, {namespace: ns})).toString() + paramsForKey(params);
```
maybe update the comment to say calling functions should not assume a valid URL is being returned, the only guarantee is that it is a unique key

benjaminapetersen replied:
Will do

benjaminapetersen commented:
Added `_isCached`, then rename `_isResourceCached` to `_isImmutable`.

Thoughts?  @jwforres

jwforres commented on assets/app/scripts/services/data.js:
```diff
@@ -1194,12 +1202,19 @@ DataService.prototype.createStream = function(resource, name, context, opts, isR
   };
 
-  var CACHED_RESOURCE = {
+  // - *should* be hit frequently enough that the LRU does not kick out as we wont fetch again
```
if it did kicked out of the cache then we'd watch it again anyway since hasImmutable checks if we have the data in the cache, maybe just update the comment so its clear, as it reads now it sounds like we are just hoping it never gets cleared out :)

benjaminapetersen replied:
Will do. I think when I wrote the comment I was in a mindset of *hoping* :smile:

jwforres commented:
just the one comment, otherwise i'm good with the updates.  needs a rebase and definitely needs a fork_ami

benjaminapetersen commented:
`fork_ami` is up and running as of last night.
Will also rebase.

benjaminapetersen commented:
Updated again.
@jhadvig, had some conflicts with DataService.js when I rebased. I think it was just the updated `.batch` method, but feel free to take a look and let me know if I missed something, my PR touches a lot of code here.

benjaminapetersen commented:
fork_ami is good.
rebasing.

benjaminapetersen commented:
rebased.
[TEST]

jwforres commented:
[merge]
